zebROS_ws/src/candle_controller/src/candle_controller.cpp

```diff
@@ -0,0 +1,255 @@
+        this->brightness_service = controller_nh.advertiseService("brightness", &CANdleController::brightnessCallback, this);
+        this->animation_service = controller_nh.advertiseService("animation", &CANdleController::animationCallback, this);
+        return true;
+    }
+
+    void starting(const ros::Time&) override {}
+
+    void update(const ros::Time&, const ros::Duration&) override {
+        const LEDs leds = *(this->led_buffer.readFromRT());
+        const Animation animation = *(this->animation_buffer.readFromRT());
+        const Brightness brightness = *(this->brightness_buffer.readFromRT());
+
+        // Brightness isn't exclusive to colours/animations, so it gets special treatment
+        if (brightness.time > this->last_write) {
+            ROS_INFO_STREAM("Writing new brightness to CANdle");
+            this->candle_handle->setBrightness(brightness.brightness);
+        }
+
+        if (leds.time > this->last_write || animation.time > this->last_write) {
+            if (leds.time > animation.time) {
+                ROS_INFO_STREAM("Writing LED group to CANdle");
+                this->last_write = leds.time;
+                this->candle_handle->setLEDGroup(leds.group);
+            } else {
+                ROS_INFO_STREAM("Writing new animation to CANdle");
+                this->last_write = animation.time;
+                this->candle_handle->setAnimation(animation.animation);
+            }
+        }
+
+        // Only overwrite last_write after the leds and animations have applied, since brightness isn't exclusive to them
+        if (brightness.time > this->last_write) {
+            this->last_write = brightness.time;
+        }
+    }
+
+    void stopping(const ros::Time&) override {}
+
+private:
+    struct LEDs {
+        ros::Time time;
+        LEDGroup group;
+
+        LEDs() {}
+    };
+
+    struct Animation {
+        CANdleAnimation animation;
+        ros::Time time;
+
+        Animation() {}
+    };
+
+    struct Brightness {
+        double brightness;
+        ros::Time time;
+
+        Brightness() {}
+    };
+
+    CANdleCommandHandle candle_handle;
+    ros::ServiceServer colour_service;
+    ros::ServiceServer brightness_service;
+    ros::ServiceServer animation_service;
+    ros::Time last_write;
+
+    realtime_tools::RealtimeBuffer<LEDs> led_buffer;
+    realtime_tools::RealtimeBuffer<Animation> animation_buffer;
+    realtime_tools::RealtimeBuffer<Brightness> brightness_buffer;
+
+    bool colourCallback(candle_controller_msgs::Colour::Request& req, candle_controller_msgs::Colour::Response&) {
+        if (this->isRunning()) {
+            LEDs leds;
+            leds.group = LEDGroup(req.start, req.count, req.red, req.green, req.blue, req.white);
+            leds.time = ros::Time::now();
+
+            this->led_buffer.writeFromNonRT(leds);
+
+            return true;
+        } else {
```

Comment:
Stylistic option nitpicking - if code returns from an if block, the else can be removed and the code in the else block made unconditional.

I have no strong opinions which is best, doesn't matter functionally.

Reply:
I'm going to leave this as-is - I find it easier to read the `else` block than assuming it from the return statement.